eclair-core/src/main/scala/fr/acinq/eclair/payment/relay/NodeRelay.scala

```diff
@@ -0,0 +1,347 @@
+    Behaviors.receiveMessage {
+      case WrappedNodeRelayPacket(IncomingPacket.NodeRelayPacket(add, outer, _, _)) => outer.paymentSecret match {
+        case None =>
+          context.log.warn("rejecting htlcId={}: missing payment secret", add.id)
+          rejectHtlc(add.id, add.channelId, add.amountMsat)
+          Behaviors.same
+        case Some(incomingSecret) if incomingSecret != secret =>
+          context.log.warn("rejecting htlcId={}: payment secret doesn't match other HTLCs in the set", add.id)
+          rejectHtlc(add.id, add.channelId, add.amountMsat)
+          Behaviors.same
+        case Some(incomingSecret) if incomingSecret == secret =>
+          context.log.debug("forwarding incoming htlc to the payment FSM")
+          handler ! MultiPartPaymentFSM.HtlcPart(outer.totalAmount, add)
+          receiving(htlcs :+ add, secret, nextPayload, nextPacket, handler)
+      }
+      case WrappedMultiPartPaymentFailed(MultiPartPaymentFSM.MultiPartPaymentFailed(_, failure, parts)) =>
+        context.log.warn("could not relay payment (paidAmount={} failure={})", parts.map(_.amount).sum, failure)
+        Metrics.recordPaymentRelayFailed(failure.getClass.getSimpleName, Tags.RelayType.Trampoline)
+        handler ! PoisonPill
+        parts.collect { case p: MultiPartPaymentFSM.HtlcPart => rejectHtlc(p.htlc.id, p.htlc.channelId, p.amount, Some(failure)) }
+        Behaviors.stopped
+      case WrappedMultiPartPaymentSucceeded(MultiPartPaymentFSM.MultiPartPaymentSucceeded(_, parts)) =>
+        val upstream = Upstream.Trampoline(htlcs)
+        handler ! PoisonPill
+        validateRelay(nodeParams, upstream, nextPayload) match {
+          case Some(failure) =>
+            context.log.warn(s"rejecting trampoline payment (amountIn=${upstream.amountIn} expiryIn=${upstream.expiryIn} amountOut=${nextPayload.amountToForward} expiryOut=${nextPayload.outgoingCltv} htlcCount=${parts.length} reason=$failure)")
+            rejectPayment(upstream, Some(failure))
+            Behaviors.stopped
+          case None =>
+            context.log.info(s"relaying trampoline payment (amountIn=${upstream.amountIn} expiryIn=${upstream.expiryIn} amountOut=${nextPayload.amountToForward} expiryOut=${nextPayload.outgoingCltv} htlcCount=${upstream.adds.length})")
+            val payFSM = relay(upstream, nextPayload, nextPacket)
+            sending(payFSM, upstream, nextPayload, fulfilledUpstream = false)
+        }
+      case _ => Behaviors.unhandled
+    }
+
+  /**
+   * Once the payment is forwarded, we're waiting for fail/fulfill responses from downstream nodes.
+   *
+   * @param upstream          complete HTLC set received.
+   * @param nextPayload       relay instructions.
+   * @param fulfilledUpstream true if we already fulfilled the payment upstream.
+   */
+  private def sending(payFSM: ActorRef, upstream: Upstream.Trampoline, nextPayload: Onion.NodeRelayPayload, fulfilledUpstream: Boolean): Behavior[Command] =
```

Comment:
I don't think you need the `payFSM` here, do you?

Reply:
We do?

Reply:
I don't see where :thinking:? Here it seems that you just pass it through, but you never actually use it to send it messages, so you don't need to keep the `ActorRef` around, do you?

Reply:
Right! Fixed as part of 43c6ca216.